rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/action/TbDeviceStateNode.java

```diff
@@ -0,0 +1,194 @@
+                    ctx.getTenantId().getId(), ctx.getSelfId().getId(), originator.getId());
+            ctx.ack(msg);
+            return;
+        }
+        if (!EntityType.DEVICE.equals(originator.getEntityType())) {
+            ctx.tellSuccess(msg);
+            return;
+        }
+        switch (event) {
+            case CONNECT_EVENT: {
+                sendDeviceConnectMsg(ctx, msg);
+                break;
+            }
+            case ACTIVITY_EVENT: {
+                sendDeviceActivityMsg(ctx, msg);
+                break;
+            }
+            case DISCONNECT_EVENT: {
+                sendDeviceDisconnectMsg(ctx, msg);
+                break;
+            }
+            case INACTIVITY_EVENT: {
+                sendDeviceInactivityMsg(ctx, msg);
+                break;
+            }
+            default: {
+                ctx.tellFailure(msg, new IllegalStateException("Configured event [" + event + "] is not supported!"));
+            }
+        }
+    }
+
+    private void sendDeviceConnectMsg(TbContext ctx, TbMsg msg) {
+        var tenantUuid = ctx.getTenantId().getId();
+        var deviceUuid = msg.getOriginator().getId();
+        var deviceConnectMsg = TransportProtos.DeviceConnectProto.newBuilder()
+                .setTenantIdMSB(tenantUuid.getMostSignificantBits())
+                .setTenantIdLSB(tenantUuid.getLeastSignificantBits())
+                .setDeviceIdMSB(deviceUuid.getMostSignificantBits())
+                .setDeviceIdLSB(deviceUuid.getLeastSignificantBits())
+                .setLastConnectTime(msg.getMetaDataTs())
+                .build();
+        var toCoreMsg = TransportProtos.ToCoreMsg.newBuilder()
+                .setDeviceConnectMsg(deviceConnectMsg)
+                .build();
+        ctx.getClusterService().pushMsgToCore(
+                ctx.getTenantId(), msg.getOriginator(), toCoreMsg, getMsgEnqueuedCallback(ctx, msg)
+        );
+    }
+
+    private void sendDeviceActivityMsg(TbContext ctx, TbMsg msg) {
+        var tenantUuid = ctx.getTenantId().getId();
+        var deviceUuid = msg.getOriginator().getId();
+        var deviceActivityMsg = TransportProtos.DeviceActivityProto.newBuilder()
+                .setTenantIdMSB(tenantUuid.getMostSignificantBits())
+                .setTenantIdLSB(tenantUuid.getLeastSignificantBits())
+                .setDeviceIdMSB(deviceUuid.getMostSignificantBits())
+                .setDeviceIdLSB(deviceUuid.getLeastSignificantBits())
+                .setLastActivityTime(msg.getMetaDataTs())
+                .build();
+        var toCoreMsg = TransportProtos.ToCoreMsg.newBuilder()
+                .setDeviceActivityMsg(deviceActivityMsg)
+                .build();
+        ctx.getClusterService().pushMsgToCore(
+                ctx.getTenantId(), msg.getOriginator(), toCoreMsg, getMsgEnqueuedCallback(ctx, msg)
+        );
+    }
+
+    private void sendDeviceDisconnectMsg(TbContext ctx, TbMsg msg) {
```

Comment:
Please move this code to the TbClusterService, as it may be used by other nodes as well. Also, we need a performance improvement to avoid pushing via the Kafka, if we run in a monolith mode.

Reply:
See DefaultSubscriptionManagerService <-> DefaultTbLocalSubscriptionService for inspiration. Maybe TbClusterService is not the best place. You may create separate service if needed that will be working in both TbCore, TbRuleEngine, and Monolith.

Reply:
Please review my implementation of this in https://github.com/thingsboard/thingsboard/pull/9030/commits/eadcf916f5c57b938776a971797f6c0d416040c8